### ISSUE TITLE
[Hackathon] Investigate options for publishing single-file binaries from NodeJS

### ISSUE THREAD

**Issue by infin8x** [pulumi/pulumi#7009]
As part of improving the authoring experience for multi-language component Pulumi Packages in #6804, we should investigate if there is a good solution in NodeJS for generating a single-file binary with no dependencies (e.g. no requirement to install NodeJS). Potential candidates:

- [nexe](https://github.com/nexe/nexe)
- [pkg](https://github.com/vercel/pkg)

**Comment by komalali** [comment]
I spent some time digging into the two suggested packages here, `pkg` and `nexe`, along with [ncc](https://www.npmjs.com/package/@vercel/ncc), a tool for compiling nodejs programs to a single file.

Some notes are captured over at: https://github.com/pulumi/pulumi/issues/7182#issuecomment-852745345

My general method was that I first ran `ncc` to compile the typescript down to 1 file, and then ran either `pkg` or `nexe`. There are some issues with `ncc` and the Pulumi SDK's dependency on `typescript` (captured [here](https://github.com/pulumi/pulumi/issues/7184)) so I ran tests that both include and exclude typescript from the resulting file. 

In summary, I tried with `pkg` for a while but it would not cooperate with some of the things that we do in the pulumi SDK, especially around runtime code. The binary that `pkg` created (70MB uncompressed, 40MB compressed) was smaller than that created by `nexe` (115MB with typescript, 90MB without), but it also didn't work (sad trombone). 

In comparison, `nexe` was much faster to do the actual compilation, but the resulting binaries are larger. Even with this drawback, overall plugin size shrunk by 60-70% in term of size, and went from 1000s of files with an external dependency on nodejs to a self-contained single executable with no external dependencies.

All in all, definitely a huge improvement on the existing state of the world. We will probably want to use a similar mechanism for storing/distributing published plugins to how we currently do for our custom resource providers, but we should spend a little more time thinking through that exact process.
